### PR TITLE
Kernel: Increase OOM safety when inserting/deleting devices, Prevent confusion of block and character devices

### PR DESCRIPTION
**Pull request by supercomputer7**

### CODE REVIEW

File: Kernel/FileSystem/DevPtsFS.cpp [hunk 1]
```diff
@@ -59,10 +59,12 @@ ErrorOr<NonnullRefPtr<Inode>> DevPtsFS::get_inode(InodeIdentifier inode_id) cons
         return *m_root_inode;
 
     unsigned pty_index = inode_index_to_pty_index(inode_id.index());
-    auto* device = DeviceManagement::the().get_device(201, pty_index);
+    auto device = DeviceManagement::the().get_device(false, 201, pty_index);
     VERIFY(device);
+    // FIXME: Find a better way to cast here
+    RefPtr<SlavePTY> pty_device = static_cast<SlavePTY*>(device.ptr());
```
Comment (BertalanD) [inline Kernel/FileSystem/DevPtsFS.cpp:65]: Do you mean something like this?
```suggestion
    auto pty_device = static_ptr_cast<SlavePTY>(device);
```

Comment (supercomputer7) [reply Kernel/FileSystem/DevPtsFS.cpp:65]: Yes. Thank you :)

**Comment by supercomputer7** [comment]
Rebased and force-pushed to solve conflicts after #11385.

**Comment by awesomekling** [comment]
> It's possible to have a device with the same major and minor numbers in a block device and character device variants.
In what situation does this happen?

**Comment by supercomputer7** [comment]
> > It's possible to have a device with the same major and minor numbers in a block device and character device variants.
> 
> In what situation does this happen?

Currently we don't have a block device and a character device with the same major and minor numbers.

However, on other systems it's entirely possible to encounter such problem. On Serenity, the real problem is that there's no real difference between character devices to block devices... the distinction is semantically only (also, some unix systems tend to define character devices vs block devices in many ways). I grasped the distinction in Serenity as "block devices for representing devices which are organized in blocks (e.g. hard drives), character devices for  devices which are working in a packet-like fashion (e.g. mouse and keyboard, RNG, ttys)", but some systems use block devices to interact with IO buffer/cache subsystem that represent real hardware underneath, and character devices for raw access.

It just seems better to prevent a situation that someone will create a block device (or a character device) that already has the same major number and minor number for an existing character device/block device.
Honestly, I'm not eager to merge this if it seems not required. I still want to ensure we can insert and remove devices safely from the kernel, but putting a distinction in the search is not too much important.

**Comment by supercomputer7** [comment]
> > > > It's possible to have a device with the same major and minor numbers in a block device and character device variants.
> > > 
> > > 
> > > In what situation does this happen?
> > 
> > 
> > Currently we don't have a block device and a character device with the same major and minor numbers.
> 
> ..
> 
> > However, on other systems ..
> 
> > It just seems better ..
> 
> I think there may be an answer here.
> 
> @supercomputer7 (edited) As long as you are willing to keep your `kernel-devices` branch, the patch will be available if needed in the future.
> 

No, I want to merge this because I want to ensure correctness and because we support both block devices and char devices, so it doesn't make sense to wait until someone merges a code that has a conflict with the given major and minor numbers and then to patch it. However, if there's another way to prevent it, I'd use that instead. I just don't think there's such option.

> BTW your PR _name_ is going to be the main _72 character commit message_, _AND_ you dont have description which will be `...` of the commit message when merged (like you did on you own commits).

When this is merged, the commit name is not going to be the PR _name_, so unless I misunderstood something, I don't know what you are talking about.

**Pull request closed without merging**